### ISSUE TITLE
Enable compositional approach of custom migration functionality

### ISSUE THREAD

**Issue by taori** [aspnet/EntityFrameworkCore#17740]
# Problem
I just read this article : https://docs.microsoft.com/de-de/ef/core/managing-schemas/migrations/operations

Reading the details about how to implement the generator objects for custom migration functionality it appears to me, that you can only extend migration functionality by overwriting existing one.

```cs
protected override void OnConfiguring(DbContextOptionsBuilder options)
    => options
        .UseSqlServer(connectionString)
        .ReplaceService<IMigrationsSqlGenerator, MyMigrationsSqlGenerator>();
```

# Solution
If a compositional approach of this were to exist it would be possible to write many packages which extend the functionality of the MigrationBuilder, which seems like a much better alternative?

**Comment by bricelam** [comment]
I like it. We have a few other places where we enable composition using services called *plugin*s. This seems to be a good fit for the same pattern.

**Comment by taori** [comment]
How would this usually be implemented though in this case? 

Overwriting the generator completely and providing this feature in an extensible way instead would always in some fashion break user code in order to work, no? 

The possibilities it opens up are endless though.

**Comment by bricelam** [comment]
We’d add a new service type like IMigrationsSqlGeneratorPlugin. You could add multiple of these. If the main generator didn’t handle an operation (or maybe before it tried to) it would see if there was a plug-in that could

**Comment by AndriySvyryd** [comment]
Also consider this for #6546

**Comment by John-Paul-R** [comment]
@roji @AndriySvyryd

I'm considering an attempt at redesigning the model differ and csharp/sql code generators to follow such a plugin system.

My personal motivation is that I am attempting to make my own EF codegen tools more reusable and pluggable -even publish them into things like an [EFCore.ModelExtras](https://github.com/John-Paul-R/EFCore.ModelExtras)- but the current layout of these classes, where one must replace each class with a single derived type makes it impossible to have more than one particular customization in an app at a time.

I'm just writing here to verify that such a contribution might be welcome, and if so to collect any insights ya'll might have or wise words on things to avoid.

Currently I'm imagining that a collection of, say, model differ implementations could be provided at startup, each optionally providing any number of operations. (may need some mechanism to override or dedup (schematized op ids? - source object references?))

Do let me know if this is something you'd like to see or discuss.

------------------------------------------------------------------------

<details>
<summary>
For context, here (expand me) is another comment I was going to send under #10770 about adding triggers and functions to EF directly, when I realized, similar to #6546, specialized code per-provider might also be useful (and also, in building the demo package linked above, I discovered the "you can only pull this trick once" predicament)
</summary>

<br/>

I'm curious if there would still be interest in an effort to enable EF Core's model to represent and migrate database triggers and functions.

For my own use, I've developed some tooling around this using the `ModelDiffer` and `CSharpCodeGenerator` extension points, and I'd be willing to try to rework this to be more EF-native/batteries-included if there's support for the general idea.

The current usage looks something like:

``` cs
public static class MyFunctions
{
    public static MyPgTriggerFunctions Trigger { get; } = new();
}

public class MyPgTriggerFunctions
{
    public static void CreateModel(ModelBuilder modelBuilder)
    {
        modelBuilder
            .DeclareFunction(LogUserEmailChange)
        ;
    }

    public readonly FunctionDeclaration LogUserEmailChange = new(
        "log_user_email_change",
        /*language=sql*/"""
        CREATE OR REPLACE FUNCTION log_user_email_change()
          RETURNS trigger
          LANGUAGE plpgsql
        AS $function$
            -- trigger function content
        $function$
        """
    );
}
// Define the functions that exist somewhere statically, to be
// referenced by other code, like "this trigger invokes X function"
// -- this let's us leverage C# type features for rudamentary reference
// checks and function discovery (a la EF.Functions).
```

``` cs
modelBuilder.Entity<User>(eb => {
    eb.HasTrigger(
        "tu_user_email_audit",
        PgTriggerTiming.After,
        [
            PgTriggerEventClause.Insert(),
            PgTriggerEventClause.Update("email"),
        ],
        PgTriggerExecuteFor.EachRow,
        MyFunctions.Trigger.LogUserEmailChange
    );
});
// Trigger creation is completely type-safe (barring the 'update' expression,
// which is a semi-arbitrary sql expression)
```

I do expect some parts of this high-level `ModelBuilder` API might have to be tailored to individual db providers (you'll note I've personally designed this for `Pg`), but I expect the backing metadata system could be pretty shared.

I've personally found the process of working with database Triggers and Functions substantially easier since switching to this system. Database logic is no longer shunted into one of many dozens of migration files (in which you have to wonder if you're looking at the latest version of the procedure). Instead, all the database procedural code lives in a dedicated space, and is versioned with git, not exclusively with a series of files. When you make a change to the procedure definition, it appears in your next migration automatically.

If you find the general idea of tracking and migrating functions and triggers within the EF Core or Npgsql appropriate, let me know, and I can put up \[do I want to put up a draft package at this point or offer to put up a PR or proposal now instead -- I might do the package either way?\] **(Edit from future me: I did do the demo package, it's here: [EFCore.ModelExtras](https://github.com/John-Paul-R/EFCore.ModelExtras))**

</details>

Edit: completing a reference loop, the [discord thread that led me to this issue](https://discord.com/channels/143867839282020352/1437297549227720795/1437297549227720795).

**Comment by roji** [comment]
@John-Paul-R please don't start any sort of serious work on EF internals without first submitting a clear design proposal which we'd discuss; otherwise you may spending time working on an approach that we'd eventually not accept.

I personally feel like before we discuss doing anything here, we need a clear, persuasive set of user scenarios which are currently impossible to support. I don't think I've seen users actually reporting use-cases where they'd need to mix multiple model differs, but I may be missing something; note the single vote that this issue has received since its filing in 2019. So as things stand, I wouldn't want you (or the EF team) to spend time on a potentially big refactoring/change for something that almost nobody needs.

**Comment by John-Paul-R** [comment]
@roji

Alright, I'll prove out as much of this as I can with an extension in a separate repo, that way I can continue to use it for my own purposes if the concept is rejected. (I think a lot of it can be done by replacing the core differ/code-generators with ones intended to be composable, and allowing plugins to hook into that)

---

As I see it, the scenario for supporting composable/pluggable SQL and C# CodeGenerators -and potentially for a similar model differ (though I believe you can get around needing to extend the ModelDiffer if you're sufficiently creative with annotations)- is really "multiple packages want to extend codegen", so this is quite core to extensibility.

Some specific discrete examples of where people might want to extend EF codegen are:

- Community authors wanting to build plugins (without having them conflict with one-another) like:
  - Support code-first setup of additional database object types, as desired in the 26-vote #10770. 
	  - _(though, as previously mentioned, I'd like to see Triggers and Functions added to EF Core outright, and would be willing to try, myself)_
	  - this does not have to stop at triggers and functions and can be made to extend to any declarative database object, like:
	    - [PgCron][pg_cron] job definitions
	    - database Views (23 + 6 vote #14537)
	    - roles
	    - replication slots
  - Skip Designer file generation 
    - (Designer files encode the whole model and therefore can be huge for large projects. They can cause project build times to skyrocket after breaking into the hundreds, and are generally not required if `dotnet ef migration remove` operations are not being used (git can be used for that))
  - other changes to the default migration file template
    - _(e.g. I would have used this for #25223 a while back)_
- Providers wanting to generate clearer migration code instead of annotation-stacking, as described in #6546
  - (right now, I get "what are these? did something break?" on >1/2 the migrations I add that add enums from my teammates because of the noisy migration annotations output


[pg_cron]: https://github.com/citusdata/pg_cron

**Comment by roji** [comment]
> Community authors wanting to build plugins (triggers, roles, replication slots...)

I sympathize with this, and I can indeed see some case for EF being better for extensibility around this. However, the problem with this is that full end-to-end support for a new database object requires doing changes in many different areas of EF:
 
* Builder APIs (which call into lower-level metadata APIs) to allow the user to e.g. configure the database object. This can in theory already be done today (doesn't require new EF extensibility APIs).
* Flow the annotations to the migrations SQL generator via the provider's IRelationalAnnotationProvider (not currently extensible).
* Rethink IMigrationsSqlGenerator to allow for some sort of plugin/extensibility API; this would have to be done at many levels (at the database for top-level objects like roles/triggers, at the table level for table customizations, at the column for column customizations...). This is non-trivial.
* Reverse-engineering (detect the thing in existing databases and scaffold the correct model configuration code for it)

Overall, I'm not sure we see big enough demand for these things (26 votes isn't very much if you look at EF's issues ordered by votes). Also, ultimately - as we usually tell people asking for various migration features - the value here simply doesn't seem high enough for doing all this work, given that users can already e.g. define roles by including SQL in their migrations. I understand the desire for everything to be handled cleanly via the model and not do SQL at all, but at the end of the day full coverage of all possible migration features across all databases simply isn't feasible, and using SQL is migrations is not something we can - or should - try to completely remove. So part of this is about making users understand that it's generally OK to use SQL in migrations, rather than expecting everything to be doable via model configuration.

> Skip Designer file generation

The migration model snapshot is a fundamental, architectural design of EF's migration system, and not something that can be "skipped" from some plugin. The dependencies on the model snapshot go beyond just `remove` - there are various migrations operation which actually require consulting the model when they're generated - these would be broken if the model snapshot is simply removed (check out e.g. SqlServerMigrationsSqlGenerator for usages of the model).

Don't get me wrong - I definitely know that the model snapshots are problematic and bloat project size, and I think there's definitely room for thinking about an alternate design. But this is not something I see as a reason for a "plugin": if we go in this direction, we'd want to come up with a new design, rather than just stopping to generate them (and break various migration features). Also, the model snapshot is required from all kinds of places - there's not a single extensibility point that would suffice here (nor would these extensibility points be necessarily relevant for the other points raised here).

> other changes to the default migration file template

I don't think we've seen a lot of need for customizability of the generate migration code; migration code isn't really part of the application, and shouldn't be interesting in any way. Practically all requests we've seen have been due to warnings being emitted for the generated code because of various formatting/code style incompatibilities with the project's; these can be simply suppressed at the migration directory level via .editorconfig or similar.

This is very different from e.g. scaffolded code which EF generates when reverse-engineering an existing database, which is meant to be used and modified in the user's project; we received many requests for customizability there, and so introduced a templating mechanism which allows the user to do what they want. We could technically do this for the migration files, but that's very different from plugin extensibility/composability (it's generally not easy to design composability for source templating that's beyond simple composability of different blocks).

> Providers wanting to generate clearer migration code instead of annotation-stacking

This again seems to be a missing feature in EF that make sense as "extensibility": it's something that needs to be implemented in EF itself and its various providers.

---

To summarize, I think some of the use-cases raised above aren't quite valid. Most shortcomings or potential feature in migrations can't be tackled simply by saying "let's introduce extensibility": they involve core EF changes that need to be done in the product (and/or in providers), and can't be done cleanly from an extensibility hook.

Otherwise, many things that do make sense via extensibility require completely different kinds of extensibility points (C# generated code customization, IMigrationSqlGenerator customization...), to the point where a general "enable compositional approach of custom migration functionality" is simply too abstract/general. So I'd really try to focus on specific user-facing problems, and a very specific extensibility point that would address those problems.

**Comment by John-Paul-R** [comment]
Hey, firstly, thanks for the quick and detailed series of replies. I greatly appreciate it!

> given that users can already e.g. define roles by including SQL in their migrations. I understand the desire for everything to be handled cleanly via the model and not do SQL at all, but at the end of the day **full coverage of all possible migration features across all databases simply isn't feasible, and using SQL is migrations is not something we can - or should - try to completely remove**. So part of this is about making users understand that it's generally OK to use SQL in migrations, rather than expecting everything to be doable via model configuration.

On this, I'd just like to clarify that I *completely* agree with this sentiment and it is actually at the heart of what I want to build here.

I would argue that the two ends of "it's generally OK to use SQL in migrations, rather than expecting everything to be doable via model configuration" are not necessarily at odds, though.

It *is* one of my goals for the model to be able to represent any object type in a database, but it is *not* one of my goals for everything to be representable without writing SQL.

(one could imagine this most trivially being achieved by just having a `ModelBuilder` method in the form below)

``` csharp
modelBuilder.HasRawObject(
   id: "pg_cron_users_rematerialzie",
   payload: """
       SELECT cron.schedule(
           'pg_cron_users_rematerialzie',
           '0 10 * * *',
           'REFRESH MATERIALIZED ...;'
       );
       """,
   teardown?: "SELECT cron.unschedule('pg_cron_users_rematerialzie');"
);
```

Dodging SQL can be nice, and I'd like to _enable_ it where feasible, but I agree that it is an impractical absolute aim. Allowing the EF model to more completely, well, model the database, has much more real benefits in my experience. Being able to just modify database functions like any other code file in my codebase is magical, and observably lowers the barrier to entry for other developers substantially, and improves discoverability massively.

(though, I do take your point that (I think) you'd consider this a vote in favor of the specific "track functions and triggers natively" feature, rather than in favor of an "extensibility" feature)

------------------------------------------------------------------------

> > Skip Designer file generation

> The migration model snapshot is a fundamental, architectural design of EF's migration system, ... (check out e.g. SqlServerMigrationsSqlGenerator for usages of the model).

Good call, I had a [mistaken assumption](https://gist.github.com/John-Paul-R/cfc1b87d164350e1058526cab2dc6a92) here. That's quite the pickle.

Thanks for the explanation, I learned something!

------------------------------------------------------------------------

> other changes to the default migration file template

> I don't think we've seen a lot of need for customizability of the generate migration code; migration code isn't really part of the application, and shouldn't be interesting in any way. ...

I agree with your points. This is definitely the least interesting/motivating region here.

------------------------------------------------------------------------

> So I'd really try to focus on specific user-facing problems, and a very specific extensibility point that would address those problems.

Would this https://github.com/John-Paul-R/EFCore.ModelExtras#getting-started type of API happen to be more in the vein of something you'd consider useful? The functions/triggers realm is what got me sent down this rabbithole. I'd be happy to focus on that if there's more immediate interest.

Barring feedback on ^, my current personal plans:

I will try to prove some of this out on my own and report back with findings. Some of these extension points feel solvable in simple enough fashion to me, particularly around `IRelationalAnnotationProvider`, `IMigrationsSqlGenerator`, and `CSharpMigrationOperationGenerator` (I think this _might_ even be possible while maintaining compatibility with existing generators as they're currently written?), but it may just be an issue of me not thinking about it completely enough, and I'll have to trip over the problem while attempting to build it. (I do agree with your point that composable text generation is non-trivial, and even less trivial to make comprehensible to consumers).

For what it's worth, my particular goal for myself in this exploration will be to enable arbitrary third-party-specified database object types from any number of discrete providers, with their corresponding migration operations and codegen, while reshuffling as few existing components as possible.

**Comment by John-Paul-R** [comment]
P.S. In case it's not clear from tone, just want to clarify that I don't expect whatever I build out to just "make it in to EF Core" if it "works" by my own standards -- I'm exploring this because it's useful and interesting to me. It's a hope and close-second goal that any solution be useful to the overall project, but at the same time I completely understand you must be judicious with what's acceptable complexity or maintenance effort on your end.

Also, thank you again for the engagement on this thread!
